### PR TITLE
Add `<Badge>` component

### PR DESCRIPTION
**Pull request by kevinzunigacuellar**
#### Description

- Adds an `<Badge>` component to @astrojs/starlight/components module.

### Notes

- Move outline style to be applied with css under `current-aria="page"`

### CODE REVIEW

**Comment by HiDeoo** [comment]
Kevin and badges, as a huge badge consumer, I'm always happy to see your PRs and this one is definitely gonna remove a lot of duplicated code for me :raised_hands:

I did not yet closely look at the code, but some general (and personal) thoughts:

- Overall, I find the badges/font a bit too big outside of the sidebar. I usually find the Vitepress version [too small](https://vitepress.dev/reference/default-theme-badge) but I think the current version is the opposite and wonder if we could find a middle ground. Note that I'm usually not the best person to ask for design feedback, so take this with a grain of salt.
- Would it make sense to also provide the `outline` variant to users?
- I wonder if allowing a `class` prop would be nice to allow users to customize the badge even further. To explain, some of the use cases I encountered where a custom class would help:
  - Defining more custom colors, for example in the requests/OpenAPI land, there are more request types than variants we have, so being able to define a custom color would be nice.
  - Having badges themed to a brand color for a company or project.

> Is an update in the package.json required?

If I understand correctly, the `packages/starlight/components.ts` update should be enough.

**Comment by kevinzunigacuellar** [comment]
Thank you for the review @HiDeoo 💜

> Overall, I find the badges/font a bit too big outside of the sidebar. I usually find the Vitepress version [too small](https://vitepress.dev/reference/default-theme-badge) but I think the current version is the opposite and wonder if we could find a middle ground. Note that I'm usually not the best person to ask for design feedback, so take this with a grain of salt.

You are right. They do look a bit bigger, I will follow the same style as the inline code block which uses a smaller font size.

> Would it make sense to also provide the outline variant to users?

I have mixed feelings about exposing `outline`. it is not an official variant since it's only applied when the sidebar is active. But then I don't see why it couldn't be an official one.

I am not sure what the best way to go about this. I could use some guidance here.

> I wonder if allowing a class prop would be nice to allow users to customize the badge even further. To explain, some of the use cases I encountered where a custom class would help:
Defining more custom colors, for example in the requests/OpenAPI land, there are more request types than variants we have, so being able to define a custom color would be nice.
Having badges themed to a brand color for a company or project.

It would be nice to have more theme versions. A class would be a very nice addition to include more customizations. 

In terms of customizing current variants. I think the nicest DX would be to override a global variable similar to what [VitePress does](https://vitepress.dev/reference/default-theme-badge#customize-type-color). 

Then `class` could override variables locally resulting in a new badge theme.

After adding `class` prop it should the user should be able to create a new theme like this

```css
.custom-badge {
  --sl-color-bg-badge: cyan;
  --sl-color-border-badge: purple;
  --sl-color-text-badge: red;
}
```

And use it like

```jsx
<Badge text="custom" class="custom-badge" />
```

**Review by HiDeoo — COMMENTED** [review]
Sorry it took me so long to get back to this PR after my first review and also for the new updates you have been working on since then. Super appreciate your patience and hard work on this. 🙌 

I left a few new comments and suggestions on the code, nothing major. I'd also like to recap everything else remaining to be done so we can push this PR to the finish line.

- ~~Confirm that the `class` prop is a good idea (I'm personally in favor as I suggested it and have use-cases for it, but I'd like to hear more opinions)~~ I talked a bit with Chris who suggested that we could actually support all valid HTML attributes for `<span>` which I think is a great idea.
- ~~Check if the new light mode styles are something we want to include (may need to check Starlight design documents)~~ Regarding the new light mode styles, after talking with Chris, a follow-up PR would be better to discuss and implement them.
- Update the branch so that we get the new updated `docs/src/content/docs/guides/components.mdx` file
- Move the new section in `docs/src/content/docs/guides/components.mdx` just above the `Icon` section
- Remove the `docs/src/content/docs/reference/test.mdx` file

Does this sound correct to you? Do you see anything missing?

File: packages/starlight/user-components/Badge.astro [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+if (!badgeVariants.includes(variant)) {
+	throw new AstroError(
+		'Invalid `variant` prop passed to the `<Badge>` component.\n',
+		`Received: ${JSON.stringify(variant)}\n` +
+			`Expected one of ${badgeVariants.map((i) => JSON.stringify(i)).join(', ')}`
+	);
+}
+---
+
+<span class:list={['sl-badge', variant, className]}><slot /></span>
```
[the remark this answers is not in the document]
Comment (kevinzunigacuellar) [reply packages/starlight/user-components/Badge.astro:21]: I agree, after taking a look after some time I think it's better to have a text prop. My previous idea was to pass in a slot in case users would like to pass svgs or other elements but I would like to keep it simple and keep it just with text.

File: packages/starlight/user-components/Badge.astro [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+const { variant = 'default', class: className } = Astro.props;
+
+if (!badgeVariants.includes(variant)) {
+	throw new AstroError(
+		'Invalid `variant` prop passed to the `<Badge>` component.\n',
+		`Received: ${JSON.stringify(variant)}\n` +
+			`Expected one of ${badgeVariants.map((i) => JSON.stringify(i)).join(', ')}`
+	);
+}
```
[the remark this answers is not in the document]
Comment (kevinzunigacuellar) [reply packages/starlight/user-components/Badge.astro:18]: Yeah! go `props` 🥳

File: packages/starlight/user-components/Badge.astro [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	.sl-badge {
+		display: inline-block;
+		border: 1px solid var(--sl-color-border-badge);
+		border-radius: 0.25rem;
+		font-family: var(--sl-font-system-mono);
+		padding: 0.125rem 0.375rem;
+		line-height: normal;
+		color: var(--sl-color-text-badge, #fff);
```
[the remark this answers is not in the document]
Comment (kevinzunigacuellar) [reply packages/starlight/user-components/Badge.astro:83]: I do not remember 😅 but happy to remove it

File: packages/starlight/user-components/Badge.astro [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		background-color: var(--sl-color-bg-badge);
+		overflow-wrap: anywhere;
+	}
+
+	/* Sidebar overrides */
+	:global(#starlight__sidebar) .sl-badge {
```
[the remark this answers is not in the document]
Comment (kevinzunigacuellar) [reply packages/starlight/user-components/Badge.astro:89]: Yeah, it seems that `#starlight__sidebar` is mostly use for media queries happy to change it.

File: packages/starlight/user-components/Badge.astro [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	/* Markdown content overrides */
+	:global(.sl-markdown-content :not(h1, h2, h3, h4, h5, h6)) .sl-badge,
+	:global(.sl-markdown-content) > .sl-badge {
+		font-size: var(--sl-text-sm);
+	}
+
+	/* outline variant */
+	:global(a[aria-current='page']) > .sl-badge {
```
[the remark this answers is not in the document]
Comment (kevinzunigacuellar) [reply packages/starlight/user-components/Badge.astro:101]: yeah, very good observation!

**Comment by kevinzunigacuellar** [comment]
Thanks for the review @HiDeoo. I added the changes you suggested, I don't think it's missing anything else besides from doing some more testing for CSS edge cases. I will do some local testing to see what pops up.

**Comment by HiDeoo** [comment]
Additional note that this PR will close #1764 by removing the `outline` CSS class which collides with Tailwind CSS [`outline`](https://tailwindcss.com/docs/outline-style) utility adding `outline-style: solid;` and leading to a thicker outline for sidebar badges for the current page:

SCR-20240418-kfwo

**Comment by kevinzunigacuellar** [comment]
I think it would be a bit complicated to handle the vertical alignment for user because it depends in two things: 
1. the size of the badge
2. the size of the parent font. 

I think baseline is a good default, users can always pass in a custom class to tweak their designs if they need to.

Thanks again for the review @HiDeoo 💜

**Comment by HiDeoo** [comment]
Are you referring to my first point, the second point, or both?

**Comment by kevinzunigacuellar** [comment]
Second point, I don't think we should support position.

**Comment by HiDeoo** [comment]
> Second point, I don't think we should support position.

Ah perfect, thanks for the clarification. Then yeah, I think we totally agree and should only try to improve the first one :thumbsup:

**Comment by HiDeoo** [comment]
Had a bit of time to play with the heading issue people mentioned during Talking & Doc'ing. It looks like it's mostly something very visible with larger headings.

I wonder if just adding a rule close to something like this could help a bit:

```css
/* Badge in headings */
:global(.sl-markdown-content :is(h1, h2, h3, h4, h5, h6)) .sl-badge {
  vertical-align: middle;
}
```

Here is a preview before and after the change. Do you have any thoughts on this? I think this may be a bit better and less something that catches the eye.

headings

**Comment by kevinzunigacuellar** [comment]
It looks good 💜, middle in only heading might work because of the bigger font sizes.

File: packages/starlight/schemas/badge.ts [hunk 6]
```diff
@@ -4,8 +4,13 @@ const badgeSchema = () =>
 	z.object({
 		variant: z.enum(['note', 'danger', 'success', 'caution', 'tip', 'default']).default('default'),
 		text: z.string(),
+		class: z.string().optional(),
```
Comment (HiDeoo) [inline packages/starlight/schemas/badge.ts:7]: If we keep this prop here instead of `BadgeComponentSchema`, we would need to document it [here](https://starlight.astro.build/reference/configuration/#badgeconfig).

File: packages/starlight/user-components/Badge.astro [hunk 7]
```diff
@@ -0,0 +1,139 @@
+---
+import { type Badge, BadgeComponentSchema } from '../schemas/badge';
+import { parseWithFriendlyErrors } from '../utils/error-map';
+
+type Props = Badge;
```
Comment (HiDeoo) [inline packages/starlight/user-components/Badge.astro:5]: At the moment, this is equivalent to `z.output<ReturnType<typeof badgeSchema>>` but I guess we would want the props to be `z.output<typeof BadgeComponentSchema>` instead so we may need to create a type for this new component schema.

**Review by HiDeoo — COMMENTED** [review]
Amazing. I just pushed this heading changes so we can have a preview including it.

I left a few comments regarding your latest changes, very minor, but this is looking really good. I would guess the documentation may still need some work but feature-wise and UI-wise I am personally happy with it.

File: packages/starlight/schemas/badge.ts (hunk 6 above)
Comment (kevinzunigacuellar) [reply packages/starlight/schemas/badge.ts:7]: Yes, it will update the corresponding docs.

**Comment by HiDeoo** [comment]
Tried to work a bit on the documentation as I was personally not super happy with the "custom badges" section as we usually have examples which includes a demo but coming up with a concise example where the demo would be looking good in both light and dark mode is a bit tricky imo.

To do that, I rewrote some parts and sharing it here at the moment as it's only a rough draft and want to get opinions first.

````mdx
### Badges

import { Badge } from '@astrojs/starlight/components';

Use the `<Badge>` component to display small pieces of information, such as status or labels.

Pass the content you want to display to the `text` attribute of the `<Badge>` component.
Optionally, you can specify a `variant` attribute with values like `tip`, `note`, `caution`, `danger`, `success` or `default` (the default using the theme accent color) to adjust the badge's appearance.

To adjust the badge's size, you can specify a `size` attribute with values like `small` (the default), `medium`, or `large`.
You can also include other `<span>` attributes such as `class`.

```mdx title="src/content/docs/example.mdx"
import { Badge } from '@astrojs/starlight/components';

<Badge text="New" variant="tip" size="small" />
<Badge text="Deprecated" variant="caution" size="medium" />
<Badge text="Starlight" variant="note" size="large" />
```

The code above generates the following on the page:

<Badge text="New" variant="tip" size="small" />
<Badge text="Deprecated" variant="caution" size="medium" />
<Badge text="Starlight" variant="note" size="large" />
````

Altho, this made me realize some points:

- We forgot [Chris suggestion](https://github.com/withastro/starlight/pull/1530#pullrequestreview-1988899462) to support all valid HTML attributes for `<span>`.
  - For consistency, we should probably remove the `class` attribute from the schema and use the same approach as [sidebar link custom HTML attributes](http://localhost:4321/guides/sidebar/#custom-html-attributes) and use an `attrs` property.
- We can probably stop using `set:html`.

What do you think?

**Comment by HiDeoo** [comment]
Amazing, love to hear that.

> I completely forgot about that set:html, definitely changing that for a set:text.

Or just `<span class:list={['sl-badge', variant, size, className]}>{text}</span>`. Considering this is equivalent and way more common, it could potentially be more readable for others.

> I wasn't very happy with the custom badges section, too. By including attributes I think it would be easier to explain that the user may override this component however they see fit.

Yeah, that's what I tried to do in my rewritten version with the "You can also include other `<span>` attributes such as `class`." sentence. It's a pattern we have used for [Link Cards](https://starlight.astro.build/guides/components/#link-cards). Of course, it could be improved as I'm not the best at documentation but I think we are on the right track.

**Comment by HiDeoo** [comment]
Thanks for the new update :tada:

I think it's looking good, and how you were able to reduce the documentation shows that imo. One thing I noticed:

- I mentioned `attrs` as a parallel with sidebar link custom HTML attributes, so that would only apply to the sidebar (not the component) but as we decided to not support that, I think we could drop entirely the new `attrs`. This would mean that in the sidebar a user can use `class` (this is currently the case) and on the component and a user could directly use `<Badge text="New" style="…" />`.
- As we would not support all attributes for sidebar badges (only `class`), I think it would be fine to just not include attributes in the Zod schema at all, pass w/e extra attributes to the `<span>` without validating them when the user is using the component directly and have the component `Props` typed with something close to `z.input<typeof badgeSideBarSchema> & HTMLAttributes<'span'>`.
  - We basically would only validate `variant`, `text` and `class` and that's it. This would allow for quite a few simplifications in the schemas I assume.

Any thoughts?

**Comment by kevinzunigacuellar** [comment]
I see what you meant. I have applied the requested changes. Thanks for the through review @HiDeoo

File: packages/starlight/schemas/badge.ts [hunk 8]
```diff
@@ -1,10 +1,14 @@
...
-	});
+const badgeSchema = () => z.object({
+	variant: z.enum(['note', 'danger', 'success', 'caution', 'tip', 'default']).default('default'),
+	text: z.string(),
+	class: z.string().optional(),
+});
+
+export const BadgeComponentSchema = badgeSchema().extend({
+	size: z.enum(['small', 'medium', 'large']).default('small'),
+});
```
Comment (HiDeoo) [inline packages/starlight/schemas/badge.ts:11]:
```suggestion
const badgeSchema = () =>
	z.object({
		variant: z.enum(['note', 'danger', 'success', 'caution', 'tip', 'default']).default('default'),
		text: z.string(),
		class: z.string().optional(),
	});

export const BadgeComponentSchema = badgeSchema()
	.extend({
		size: z.enum(['small', 'medium', 'large']).default('small'),
	})
	.passthrough();

export type BadgeComponentProps = z.input<typeof BadgeComponentSchema>;
```

We can use [`.passthrough`](https://github.com/colinhacks/zod#passthrough) so that unknown keys are not stripped.

File: packages/starlight/user-components/Badge.astro [hunk 9]
```diff
@@ -0,0 +1,153 @@
+---
+import { BadgeComponentSchema } from '../schemas/badge';
```
Comment (HiDeoo) [inline packages/starlight/user-components/Badge.astro:2]:
```suggestion
import { BadgeComponentSchema, type BadgeComponentProps } from '../schemas/badge';
```

File: packages/starlight/user-components/Badge.astro [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+const { text, variant, size, class: customClass } = parseWithFriendlyErrors(
+	BadgeComponentSchema,
+	{
+		class: classProp,
+		variant: variantProp,
+		text: textProp,
+		size: sizeProp,
+	},
+	'Invalid prop passed to the `<Badge/>` component.'
+);
```
Comment (HiDeoo) [inline packages/starlight/user-components/Badge.astro:30]:
```suggestion
type Props = BadgeComponentProps & HTMLAttributes<'span'>;

const {
	text,
	variant,
	size,
	class: customClass,
	...attrs
} = parseWithFriendlyErrors(
	BadgeComponentSchema,
	Astro.props,
	'Invalid prop passed to the `<Badge/>` component.'
);
```

The previous change allows for a bit of simplification in the component.

**Review by HiDeoo — COMMENTED** [review]
Amazing. I left 2 small suggestions to simplify a little bit the code and avoid some duplication, but personally I'm very happy with the new component :tada:

You did an amazing work 👏

File: packages/starlight/user-components/Badge.astro (hunk 10 above)
Comment (kevinzunigacuellar) [reply packages/starlight/user-components/Badge.astro]: Thanks Hideo 💜

File: docs/src/content/docs/guides/components.mdx [hunk 11]
```diff
@@ -368,6 +368,35 @@ import { Steps } from '@astrojs/starlight/components';
...
+import { Badge } from '@astrojs/starlight/components';
+
+Use the `<Badge>` component to display small pieces of information, such as status or labels.
+
+Pass the content you want to display to the `text` attribute of the `<Badge>` component.
+Optionally, you can specify a `variant` attribute with values like `tip`, `note`, `caution`, `danger`, `success` or `default` (the default uses the theme accent color) to adjust the badge's appearance.
+
+To adjust the badge's size, you can specify a `size` attribute with values like `small` (the default), `medium`, or `large`.
+
+You can also include other `<span>` attributes such as `class` or `style` to further customize the badge.
```
Comment (sarah11918) [inline docs/src/content/docs/guides/components.mdx:382]:
```suggestion
Pass the content you want to display to the `text` attribute of the `<Badge>` component.

By default, the badge will use the theme accent color of your site. To use a built-in badge color, set the `variant` attribute to one of the following values: `note` (blue), `tip` (purple), `danger` (red), `caution` (orange), or `success` (green).

The `size` attribute (default: `small`) controls the size of the badge text. `medium` and `large` are also available options for displaying a larger badge.

For further customization, use other `<span>` attributes such as `class` or `style` with custom CSS.
```
I'm thinking something like this? Adjust to taste, or ignore!

**Review by sarah11918 — COMMENTED** [review]
Docs look great, @kevinzunigacuellar ! Just left some polishing thoughts below!

File: docs/src/content/docs/guides/sidebar.mdx [hunk 12]
```diff
@@ -294,12 +294,14 @@ The configuration above generates the following sidebar:
...
 
-### Badge variants
+### Badge variants and custom styling
 
-Customize the badge styling using an object with `text` and `variant` properties.
+Customize the badge styling using an object with `text`, `variant` and `class` properties.
 
 The `text` represents the content to display (e.g. "New").
-Override the `default` styling, which uses the accent color of your site, by setting the `variant` property to one of the following values: `note`, `tip`, `danger`, `caution` or `success`.
+Change the appearance, by setting the `variant` property to one of the following values: `note`, `tip`, `danger`, `caution` or `success`.
```
[the remark this answers is not in the document]
Comment (kevinzunigacuellar) [reply docs/src/content/docs/guides/sidebar.mdx:302]: Variant is optional and defaults to `default`. Also sizing is not available for sidebar items and groups.

File: docs/src/content/docs/reference/frontmatter.md [hunk 13]
```diff
@@ -340,7 +340,7 @@ sidebar:
 
 Add a badge to the page in the sidebar when displayed in an autogenerated group of links.
 When using a string, the badge will be displayed with a default accent color.
-Optionally, pass a [`BadgeConfig` object](/reference/configuration/#badgeconfig) with `text` and `variant` fields to customize the badge.
+Optionally, pass a [`BadgeConfig` object](/reference/configuration/#badgeconfig) with `text`, `variant` and `class` fields to customize the badge.
```
[the remark this answers is not in the document]
Comment (kevinzunigacuellar) [reply docs/src/content/docs/reference/frontmatter.md:343]: Can't change the size in the sidebar.

File: docs/src/content/docs/reference/configuration.mdx [hunk 14]
```diff
@@ -203,6 +203,7 @@ type SidebarItem = {
 interface BadgeConfig {
 	text: string;
 	variant: 'note' | 'tip' | 'caution' | 'danger' | 'success' | 'default';
```
[the remark this answers is not in the document]
Comment (kevinzunigacuellar) [reply docs/src/content/docs/reference/configuration.mdx:205]: Yes, variant is optional and size is not available.

Comment (kevinzunigacuellar) [inline docs/src/content/docs/reference/configuration.mdx:205]:
```suggestion
	variant?: 'note' | 'tip' | 'caution' | 'danger' | 'success' | 'default';
```

File: .changeset/wicked-melons-study.md [hunk 15]
```diff
@@ -0,0 +1,6 @@
+---
+'@astrojs/starlight': minor
+---
+
+Adds the `<Badge>` component
```
[the remark this answers is not in the document]
Comment (kevinzunigacuellar) [reply .changeset/wicked-melons-study.md:5]: yes, it is the same component but now is generally available.

File: docs/src/content/docs/reference/frontmatter.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (delucis) [reply docs/src/content/docs/reference/frontmatter.md:343]: Cool — guess we still should add the Oxford comma (if we’re being picky):

```suggestion
Optionally, pass a [`BadgeConfig` object](/reference/configuration/#badgeconfig) with `text`, `variant`, and `class` fields to customize the badge.
```

File: .changeset/wicked-melons-study.md [hunk 16]
```diff
@@ -0,0 +1,6 @@
+---
+'@astrojs/starlight': minor
+---
+
+Expands the `<Badge>` component to be generally available
```
Comment (delucis) [inline .changeset/wicked-melons-study.md:5]: I know we think of this as “expanding”, but I actually think we can describe this as “adding” a component — the previous component doesn’t really exist for users currently, they only enable it via config options.

```suggestion
Adds a new `<Badge>` component
```

File: docs/src/content/docs/guides/sidebar.mdx [hunk 17]
```diff
@@ -294,12 +294,14 @@ The configuration above generates the following sidebar:
 	]}
 />
 
-### Badge variants
+### Badge variants and custom styling
 
-Customize the badge styling using an object with `text` and `variant` properties.
+Customize the badge styling using an object with `text`, `variant` and `class` properties.
```
Comment (delucis) [inline docs/src/content/docs/guides/sidebar.mdx:299]: Ooh, spotted another Oxford comma opportunity 👀 
```suggestion
Customize the badge styling using an object with `text`, `variant`, and `class` properties.
```

File: packages/starlight/components/SidebarSublist.astro [hunk 18]
```diff
@@ -24,13 +24,11 @@ const { sublist, nested } = Astro.props;
...
-								<Badge
-									text={entry.badge.text}
-									variant={entry.isCurrent ? 'outline' : entry.badge.variant}
-								/>
-							</>
+							<Badge
+								variant={entry.badge.variant}
+								class={entry.badge.class}
+								text={entry.badge.text}
+							/>
```
Comment (delucis) [inline packages/starlight/components/SidebarSublist.astro:31]: It’s worth double checking that removing the fragment and explicit `{' '}` whitespace here doesn’t cause issues — it was added as a workaround for issues caused by https://github.com/withastro/compiler/issues/852

File: packages/starlight/components/SidebarSublist.astro [hunk 19]
```diff
@@ -41,10 +39,11 @@ const { sublist, nested } = Astro.props;
 							<div class="group-label">
 								<span class="large">{entry.label}</span>
 								{entry.badge && (
-									<>
-										{' '}
```
Comment (delucis) [inline packages/starlight/components/SidebarSublist.astro:45]: Same here

**Review by delucis — COMMENTED** [review]
Thanks again for all the work advocating for this component @kevinzunigacuellar! Can’t wait to release it. Left a couple of final small notes.

File: packages/starlight/components/SidebarSublist.astro (hunk 18 above)
Comment (delucis) [reply packages/starlight/components/SidebarSublist.astro:31]: I don’t see the whitespace stripped in the preview, but maybe I’m missing something?

Comment (kevinzunigacuellar) [reply packages/starlight/components/SidebarSublist.astro:31]: Tested a local build in chrome and safari, it seems to be working fine.

**Review by delucis — APPROVED** [review]
I hereby approve this PR! ✅ 

Will release it later today — thanks again for your work @kevinzunigacuellar and to @HiDeoo for shepherding it through 💖
